File: Client/Captions/CaptionsLib/source/BaseCaptionWriter.vb
'Filename: BaseCaptionWriter.vb
'Version: 20121016

Imports ClipFlair.CaptionsLib.Models

Imports System.IO
Imports System.Text
Imports Microsoft.SilverlightMediaFramework.Core.Accessibility.Captions

Namespace ClipFlair.CaptionsLib

  Public MustInherit Class BaseCaptionWriter
    Implements ICaptionsWriter

#Region "Methods"

    Public Overloads Sub WriteCaptions(ByVal captions As CaptionRegion, ByVal path As String, ByVal theEncoding As Encoding) Implements ICaptionsWriter.WriteCaptions
      Using writer As New StreamWriter(path, False, theEncoding)
        WriteCaptions(captions, writer)
      End Using
    End Sub

    Public Overloads Sub WriteCaptions(ByVal captions As CaptionRegion, ByVal stream As Stream, ByVal theEncoding As Encoding) Implements ICaptionsWriter.WriteCaptions
      Using writer As New StreamWriter(stream, theEncoding)
        WriteCaptions(captions, writer)
      End Using
    End Sub

    Public Overloads Sub WriteCaptions(ByVal captions As CaptionRegion, ByVal writer As TextWriter) Implements ICaptionsWriter.WriteCaptions
      Try
        WriteHeader(writer)
        For Each s As CaptionElement In captions.Children
          WriteCaption(s, writer)
        Next
        WriteFooter(writer)
      Finally
        writer.Close()
      End Try
    End Sub

    Public Overridable Sub WriteHeader(ByVal writer As TextWriter) Implements ICaptionsWriter.WriteHeader
      'can override at descendents
    End Sub

    Public MustOverride Sub WriteCaption(ByVal Caption As CaptionElement, ByVal writer As TextWriter) Implements ICaptionsWriter.WriteCaption

    Public Overridable Sub WriteFooter(ByVal writer As TextWriter) Implements ICaptionsWriter.WriteFooter
      'can override at descendents
    End Sub

#End Region

  End Class

End Namespace

File: Client/Captions/CaptionsLib/source/BaseCaptionWriter.cs
//Project: ClipFlair (http://ClipFlair.codeplex.com)
//Filename: BaseCaptionWriter.cs
//Version: 20131113

using ClipFlair.CaptionsLib.Utils;
using ClipFlair.CaptionsLib.Models;

using System.IO;
using System.Text;
using Microsoft.SilverlightMediaFramework.Core.Accessibility.Captions;

namespace ClipFlair.CaptionsLib
{

	public abstract class BaseCaptionWriter : ICaptionsWriter
	{

		#region --- Methods ---

		public void WriteCaptions(CaptionRegion captions, string path, Encoding theEncoding)
		{
			using (StreamWriter writer = new StreamWriter(path, false, theEncoding)) {
				//the Using statement will close the file created when finished
				WriteCaptions(captions, writer);
			}
		}

		public void WriteCaptions(CaptionRegion captions, Stream stream, Encoding theEncoding)
		{
			StreamWriter writer = new StreamWriter(stream, theEncoding);
			//not Using statement, do not close the stream when finished
			WriteCaptions(captions, writer);
		}

		public void WriteCaptions(CaptionRegion captions, TextWriter writer)
		{
			writer.NewLine =  StringUtils.vbCrLf;
			WriteHeader(writer);
			foreach (CaptionElement c in captions.Children) {
				WriteCaption(c, writer);
			}
			WriteFooter(writer);
			writer.Flush();
			//write out any buffered data
		}

		public virtual void WriteHeader(TextWriter writer)
		{
			//can override at descendents
		}

		public abstract void WriteCaption(CaptionElement caption, TextWriter writer);

		public virtual void WriteFooter(TextWriter writer)
		{
			//can override at descendents
		}

		#endregion

	}

}